### PR TITLE
Adds Graphflood v1 to libtopotoolbox

### PR DESCRIPTION
**Pull request by bgailleton**
This pull request adds a first version of Graphfloods to libtopotoolbox.

Everything is self-contained in a subfolder, the only "conflict" should happen in the cmake files.

I tested it from the `python` side - explaining the high number of commits (required at every changes to propagate compilation).

Includes:
- graphflood ("full" as described in the method section of Gailleton et al., 2024) multiple flow and single flow
- single flow graph with custom boundary conditions
- Variants of the Priority Floods algorithm to deal with local minimas and impose slopes
- some utility functions for custom neighbouring
- FIFO Queue and dedicated Priority Queue data structures

### CODE REVIEW

**Comment by wkearn** [comment]
This is great, @bgailleton! It might take a bit of time to work my way through it, but I think it will be a good thing to have in libtopotoolbox.

It looks like there are some merge conflicts in `include/topotoolbox.h` and `src/CMakeLists.txt`. Do you think you can fix those up? I guess our CI doesn't run on a PR unless the PR is conflict-free. If you have trouble with the conflicts, I can also try to sort it out: I think it might be because of some recent PRs that I merged.

**Comment by bgailleton** [comment]
@wkearn I resolved the conflicts, they were indeed due to recent PRs. Nothing really conflicting, just updating the interface (e.g. dim) and related comments. I moved all of graphflood code at the end of the file.

A few notes:
- I have not tried with COLUMN MAJOR languages
- So far I use my own flow routing as I need some specificities (e.g. fine-grain boundary conditions) but I'll be happy to update them in the future with other flow routines
- I use #pragma once as header guard (supported by all mainstream and not-so-mainstream C compilers now)
- I'll add more interfaces in the future with other PRs

Let me know if you need other infos

**Comment by bgailleton** [comment]
Just notice the CI considers all warnings as error, do I need to sort them all?

File: src/priority_queue.c [hunk 1]
```diff
@@ -74,7 +74,7 @@ PriorityQueue pq_create(ptrdiff_t max_size, ptrdiff_t *heap, ptrdiff_t *back,
 
   if (flag) {
     q.count = max_size;
-    build_heap(&q);
+    pq_deletemin(&q);
   }
```
[the remark this answers is not in the document]
Comment (wkearn) [reply src/priority_queue.c:78]: This change also causes a test failure for our `excesstopography_fmm*` functions, presumably because their priority queues are not constructed fully. Reverting it makes all our existing tests pass.

[the remark this answers is not in the document]
Comment (bgailleton) [reply src/priority_queue.c:78]: Ah no, I did not notice I changed that, I initially used your implementation of priority queue and changed to mine because I needed specific operations. So it must be a mistake

[the remark this answers is not in the document]
Comment (bgailleton) [reply src/priority_queue.c:78]: So just to be clear that was not intentional and you can reverse any change I made to any file not in `graphflood` subfolder (apart from `topotoolbox.h` or `CMakeLists.txt`)

**Comment by wkearn** [comment]
I've looked through these changes, and everything makes sense to me, though I haven't run anything yet -- that's the next goal. Here are some initial thoughts. Sorry that this got kind of long: it turned into a dump of things that I want to keep in mind. The highlights:

1. Are there any low-complexity tests that we can add to our automated testing to make sure we don't see too many regressions as we make changes going forward?
2. Is there any way that you would like to test graphflood but can't because our testing infrastructure isn't sufficient? I can prioritize improvements there if it would help.
3. Let's make sure that CI passes, which should require fixing the warnings and reformatting everything with `clang-format`. The existing tests should pass, once we've reverted that change to the priority queue.
4. I do have some questions about dynamic memory allocation, discussed below.
5. I have some minor stylistic comments that I will go through and add to specific lines. Things like argument types and order that it would be nice to keep consistent with the rest of libtopotoolbox.

If you're okay with me pushing changes to your branch, I can take care of (3). If you'd rather handle it yourself, it shouldn't be too complicated.

Now the brain dump:

# Testing

Do you have any formal tests of graphflood that you like to use beyond just running it and making sure everything looks right? I'd like to have some automated tests just to give ourselves some confidence that we haven't broken anything if we need to make changes in the future. 

Our testing infrastructure is not very sophisticated (see #55), but if there is a particular testing strategy that would work well for your stuff, we can make it happen. Right now we run property-based tests on randomly generated DEMs (see `test/random_dem.cpp` and `test/excesstopography.cpp`). At the very least, we can exercise graphflood in the same way, which can conceivably catch things like indexing errors, even if it is hard to find stringent, testable properties. The recently added TT2-style flow routing stuff is under-tested for that reason.

# Dynamic memory allocation

A major thing that I do think we need to resolve is the dynamic allocation of memory. Your usage of malloc is pretty simple: allocate memory at the start of the function and free it at the end, so I'm not too worried about use-after-free or memory leaks or other memory safety problems. 

I am, however, worried about out-of-memory errors and how we handle them. It's conceivable that a user working on a large DEM or landscape evolution model could run out of memory. At the moment, if a user calls `graphflood_full` without enough memory available, it will trigger a segfault when they try to deference a pointer that came from a malloc failure, and their application will crash without explanation.

The alternative that I usually prefer is to pass all of the needed memory into the function, which is what we do with, e.g. `gwdt`. Allocating buffers and handling out-of-memory errors becomes Python's or MATLAB's responsibility. It does make the function interfaces more complex. So far we've only required three or four intermediate arrays to be passed. I can see how passing a dozen intermediate arrays to `graphflood_full` will get unwieldy.

Another option is to build more error handling into libtopotoolbox, but I would rather not take on that complexity if we can avoid it.

In summary: I think the current situation, where out-of-memory errors will just crash is not ideal, but I'm entirely not sure what the best way to move forward is. For now, the best thing might be just to warn the user in the documentation that this can fail. Then I would feel better merging what you have now while we experiment with different approaches.

# Duplication

There is a fair amount of duplication between what we currently have and what graphflood needs. As you mentioned there are subtly different requirements that, for example, make us need different priority queues. Also something like your `compute_sfgraph` or `compute_drainage_area_single_flow` is quite similar to what I've just recently implemented in `flow_routing_d8_carve` and `flow_accumulation`.

Resolving this duplication doesn't block merging this PR, and I actually think it's a good idea to keep the duplicates for now just so we can see how different things work in practice. I would eventually like to combine things where it makes sense. If we're going to go to the trouble of maintaining our own data structures, it is probably a better idea to have one than to have several slightly different ones.

I like a lot of the `gf_utils`, and I think I will rewrite some existing libtopotoolbox stuff to use them.

**Comment by bgailleton** [comment]
Thanks for the thorough review. I'll try to answer points by points and have a couple of questions too.

Quick questions first - feel free to let me know if it does not match libtopotoolbox design and philosophy:

A) Would you be OK to merge once the warnings and `clang-format` issues are sorted + other changes you think are essential? I would then work on adding tests, other features, improvements and potential fixes via PRs when I get to dedicate time to that.
B) I am happy to maintain `graphflood` in the future and adapt it to other parts of ttb (e.g. flow routing) but how do you feel if I keep some functions exposed for specific use cases (e.g. my single flow graph for some applications that requires super specific boundary conditions?). Alternatively, I could expose them in my own `c++` backends that's not a problem.

- **1.  and 2.:** I can design some tests around drainage area continuity on random landscapes and eventually analytical solutions on simple cases. I don't think there is anything missing for minimal tests (although I won't be able to design them in the coming days)
- **3.** It might be good if I learn to format the code? just to make sure I do it for the next PRs. I can fix the warnings, could you just point me toward quick instructions for `clang-format`? I have always used it behind the scene via github actions but don't really know the tool
- **5.** Sure go ahead, I am starting to bind it to my own front-end so just tag me when it's done so I can propagate the changes.


### Dynamic Memory allocation:

On the `python` side, I get some relatively explicit warnings/error if I do not free the memory within C (I am not sure why or how but I forgot a couple of free and I figured it out like that) so I am not too worried about leaks too.
I would be reluctant to pass everything from python/MATLAB: these versions of graphflood are minimal and without optional outputs and it would already ask for ~10 arrays + some parameters, while at the moment all the non-user inputs are behind the scene. If I start adding  more sophisticated versions of graphflood and/or optional outputs (there can be a lot - e.g. velocity, discharge in/out, ...) it starts to become very long.

For memory management, I do not have a perfect solution either, note that c++ has `std::bad_alloc` that's easily catchable. Another solution would be to raise a warning from python/MATLAB that take into account the size of the DEM vs available memory so that if the program crashes the user would have been warn that the analysis might not be suited for their laptop?

### Duplication

I am happy to reduce the number of duplication in future PRs (and feel free to do it if you want/need before then). Most of them are/will be due to me not being familiar with the whole of TTB. For I actually exposed the drainage area function for testing purposes. As for the almost-duplicates and derived functions, I am happy to keep them internal or explicitely reversed for internal purposes if that is a possibility for you?

Feel free to use, copy upstream as much as `gf_utils` as needed, glad you like it. I have similar logic in all my backends to deal with generic neighbouring operations and standardised boundary conditions (makes easy to send the same boundary conditions array to my GPU stuff and then to TTB to analyse it and now the neighouring will remain consistent)

**Comment by bgailleton** [comment]
@wkearn I applied `clang-format` and fixed all the warnings. 

No tests yet but the expample from the `python` repo produces correct results, I'll add some later on.

I think it is ready to merge?

**Comment by bgailleton** [comment]
OK I forgot that other compilers would have different needs, I'll check these

**Comment by wkearn** [comment]
> OK I forgot that other compilers would have different needs, I'll check these

Yeah, that's always a little annoying. If you need some quicker feedback, you can switch on the GitHub Actions on your fork -- it looks like I need to manually approve your CI runs until you become a "contributor." You need to apply

```diff
--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -4,7 +4,7 @@ on:
   pull_request:
     branches: ["main"]
   push:
-    branches: ["main"]
+    branches: ["main","graphflood"]

 jobs:
   format-check:
```

to trigger CI runs on your fork. Then you can revert this patch once you've got everything working.

**Comment by bgailleton** [comment]
ah yes I need to format again haha I'll get this right and I'll change the workflow

**Comment by bgailleton** [comment]
@wkearn the workflow is running on my local branch, you can ignore the request for approval I will tag you when ready

**Comment by bgailleton** [comment]
SHould be ok for merging @wkearn

**Review by wkearn — APPROVED** [review]
